src/core/mac/sub_mac_wed.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ *  Copyright (c) 2024, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Wake-up End Device of the subset of IEEE 802.15.4 MAC primitives.
+ */
+
+#include "sub_mac.hpp"
+
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+
+#include "instance/instance.hpp"
+
+namespace ot {
+namespace Mac {
+
+RegisterLogModule("SubMac");
+
+void SubMac::WedInit(void)
+{
+    mWedListenInterval = 0;
+    mWedTimer.Stop();
+}
+
+void SubMac::UpdateWakeupListening(uint32_t aInterval, uint16_t aDuration, uint8_t aChannel)
+{
+    VerifyOrExit(RadioSupportsReceiveTiming());
+    VerifyOrExit(aInterval != mWedListenInterval || aDuration != mWedListenDuration || aChannel != mWakeupChannel);
+
+    mWedListenInterval = aInterval;
+    mWedListenDuration = aDuration;
+    mWakeupChannel     = aChannel;
+    mWedTimer.Stop();
+
+    if (mWedListenInterval > 0)
+    {
+        mWedSampleTime      = TimerMicro::GetNow() + kCslReceiveTimeAhead;
+        mWedSampleTimeRadio = otPlatRadioGetNow(&GetInstance()) + kCslReceiveTimeAhead;
+
+        if ((mWedSampleTime.GetValue() > mWedListenInterval) && (mWedSampleTimeRadio > mWedListenInterval))
+        {
+            mWedSampleTime -= mWedListenInterval;
+            mWedSampleTimeRadio -= mWedListenInterval;
+        }
```

Comment:
What is the reason for this `if` check? Not sure if we need these?

If the intention is to avoid underflow in `mWedSampleTime -= mWedListenInterval`, it should be unnecessary because `TimeMicro` handles overflow and underflow correctly.

Reply:
Yes, the idea here is to subtract `mWedListenInterval` from the sample times the first time since it will be recurrently added in `HandleWedTimer`.

The problem could come if the feature gets enabled before `mWedListenInterval` microseconds have passed since boot up, at least for `mWedSampleTimeRadio` which is `uint64_t` (underflow not handled). So I'll keep that one check.

Reply:
We should not need this.

It works fine with both `TimeMicro` and `uint64_t` value. Anytime we deal with any `uint` calculation, the underflow/overflow is well-defined  and uses `mod 2^n` (unlike signed `int` where behavior is not explciltly defined by C++ reference and can be compiler dependent). `TimeMicro/Milli` don't do anything special to handle underflow/overflow, they simply use `uint32_t` as the underlying type.

For instance, if `mWedSampleTimeRadio` is zero and we subtract the interval from it, the result would be  a large `0xffff...` value. Adding the same `mWedListenInterval` interval in `HandleWedTimer()` to the result will correctly restores the value back to zero.

Here is link to reference: https://en.cppreference.com/w/cpp/language/operator_arithmetic
> **Overflows**
> Unsigned integer arithmetic is always performed modulo 2n where n is the number of bits in that particular integer. E.g. for unsigned int, adding one to UINT_MAX gives ​0​, and subtracting one from ​0​ gives UINT_MAX
>
> When signed integer arithmetic operation overflows (the result does not fit in the result type), the behavior is undefined, ...

Reply:
The whole `if` block could be removed if `HandleWedTimer` increased `mWedSampleTime[Radio]` after (not before) firing the timer and calling `ReceiveAt`.

Reply:
I have just unconditionally moved the subtraction to the previous block.